lib/icingadb/icingadbchecktask.cpp

```diff
@@ -0,0 +1,386 @@
+
+	if (service)
+		resolvers.emplace_back("service", service);
+	resolvers.emplace_back("host", host);
+	resolvers.emplace_back("command", commandObj);
+	resolvers.emplace_back("icinga", IcingaApplication::GetInstance());
+
+	auto resolve ([&resolvers, &checkable, &resolvedMacros, useResolvedMacros](const String& macro, String* missing) {
+		return MacroProcessor::ResolveMacros(macro, resolvers, checkable->GetLastCheckResult(),
+			missing, MacroProcessor::EscapeCallback(), resolvedMacros, useResolvedMacros);
+	});
+
+	String icingadbName = resolve("$icingadb_name$", nullptr);
+
+	String missingDownForWarning;
+	String missingDownForCritical;
+	String missingHeartbeatWarning;
+	String missingHeartbeatCritical;
+	String missingIdleForWarning;
+	String missingIdleForCritical;
+	String missingHistoryBacklogWarning;
+	String missingHistoryBacklogCritical;
+	String missingQueriesWarning;
+	String missingQueriesCritical;
+	String missingPendingQueriesWarning;
+	String missingPendingQueriesCritical;
+
+	Value downForWarning = resolve("$icingadb_downfor_warning$", &missingDownForWarning);
+	Value downForCritical = resolve("$icingadb_downfor_critical$", &missingDownForCritical);
+	Value heartbeatWarning = resolve("$icingadb_heartbeat_warning$", &missingHeartbeatWarning);
+	Value heartbeatCritical = resolve("$icingadb_heartbeat_critical$", &missingHeartbeatCritical);
+	Value idleForWarning = resolve("$icingadb_idlefor_warning$", &missingIdleForWarning);
+	Value idleForCritical = resolve("$icingadb_idlefor_critical$", &missingIdleForCritical);
+	Value historyBacklogWarning = resolve("$icingadb_history_backlog_warning$", &missingHistoryBacklogWarning);
+	Value historyBacklogCritical = resolve("$icingadb_history_backlog_critical$", &missingHistoryBacklogCritical);
+	Value queriesWarning = resolve("$icingadb_queries_warning$", &missingQueriesWarning);
+	Value queriesCritical = resolve("$icingadb_queries_critical$", &missingQueriesCritical);
+	Value pendingQueriesWarning = resolve("$icingadb_pending_queries_warning$", &missingPendingQueriesWarning);
+	Value pendingQueriesCritical = resolve("$icingadb_pending_queries_critical$", &missingPendingQueriesCritical);
+
+	struct Threshold
+	{
+		Value Val;
+		String Missing;
+	};
+
+	struct Thresholds
+	{
+		Threshold Warning, Critical;
+	};
+
+	std::map<String, Thresholds> thresholdsByOp;
+
+	const char * const icingadbKnownStats[] = {
+		"sync_config", "sync_state", "sync_history", "sync_overdue", "cleanup_history"
+	};
+
+	struct {
+		const char * Kind;
+		Threshold Thresholds::* Field;
+	} const thresholdKinds[] = {{"warning", &Thresholds::Warning}, {"critical", &Thresholds::Critical}};
+
+	for (auto metric : icingadbKnownStats) {
+		Thresholds thresholds;
+
+		for (auto wc : thresholdKinds) {
+			auto& threshold (thresholds.*wc.Field);
+
+			threshold.Val = resolve(String("$icingadb_") + metric + "_" + wc.Kind + "$", &threshold.Missing);
+		}
+
+		thresholdsByOp.emplace(metric, std::move(thresholds));
+	}
```

Comment:
I think just doing the following is just fine. Also it's easier to understand as it avoids the really nice pointer to member syntax and it's even shorter.

```cxx
	for (auto metric : icingadbKnownStats) {
		Thresholds thresholds;

		thresholds.Warning.Val = resolve(String("$icingadb_") + metric + "_warning$", &thresholds.Warning.Missing);
		thresholds.Critical.Val = resolve(String("$icingadb_") + metric + "_critical$", &thresholds.Critical.Missing);

		thresholdsByOp.emplace(metric, std::move(thresholds));
	}
```

Plus when combined with my previous suggestion on `resolve()`, things get even nicer.